### ISSUE TITLE
Machine readable dependency graph

### ISSUE THREAD

**Issue by PelleUllberg** [flaminem/flamy#1]
Hi, Thanks for an awesome project. 

We have 100:ds of different packaged Hive transformation projects. Each project is a bunch of create and insert statements that produces projections and aggregates of other datasets. So far, each package executes all the sql in sequence. We've been discussing to add a tool that analyses the dependencies between statements and if possible creates a parallelised execution tree to speed up wall clock execution time of the package set of statements. 

I used flamy to produce a graph for one of our bigger projects, and it work great! The graph shows a lot of "plateaus". My working thesis is that all transformations on the same plateaus should be able to execute in parallell since they don't have any dependencies to each other. 

So, in order to take this further, I need a machine readable representation of the dependency graph so I can use that to materialise an oozie workflow with forks and joins.

In other words, how can I access the dependency model ni Flamy before the png is rendered?

best regards
/Per Ullberg

P.S. Check out https://github.com/klarna/HiveRunner

**Comment by FurcyPin** [comment]
Hi Pelle,

Thank you for your interest in flamy!
Flamy builds an internal dependency model as a Scala object, but I can easily add a 
new command to export it under some format, json or else, or even directly generate 
a whole oozie workflow xml file: just tell me what you need.

Also, did you try the "run" command, that is already able to execute all the required queries
from a point of the graph to another with optimal parallelism?
In our use case we don't use Oozie, but another scheduler of our own, and simply run complete
Hive workflows with a simple "flamy run" command called by our scheduler.

Please feel free to email me, I would be glad to discuss your needs more further via Skype or Hangout.

Regards,

Furcy

**Comment by FurcyPin** [comment]
ping? 

I really want to help you on this, but I need your feedback: Which format would you prefer? json, xml?

If you provide me with a minimal example of hive tables and the expected corresponding Oozie workflow, I can even add a feature to export the graph as a Oozie workflow.

If you want something quick and dirty, you can also look in the folder where the .png are generated: there are also .dot files which should be parsable, but I don't think this would be the correct way to go.

**Comment by PelleUllberg** [comment]
Sorry for being unresponsive. It's currently not on the top of the backlog,
it was part of an investigation.

We're pretty good at materializing oozing workflows, so that's not an
issue. One thing I have not looked into yet with flamy is how to inject
settings... for instance, we sometimes switch between tez and MR in the
middle of a workflow between two queries.

How's that handled?

Regards
/Pelle

**Comment by FurcyPin** [comment]
Hi,

Flamy has two ways to inject settings:
You can either have a file containing global Hive settings that will be added to each Hive query run by flamy, which is explained here: http://flamy.readthedocs.io/en/latest/FAQ.html#presets-files
This is useful to add a new default global setting without restarting the HiveServer.

If you want to apply custom Hive settings to a single query (e.g. to optimize it), you can simply add
any number of SET commands before your query in the populate file.

So for your use case, I guess all you have to do is add `SET hive.execution.engine = mr` or `SET hive.execution.engine = tez` before each query that you want to execute with MapReduce or Tez, whichever is not your default engine.

Regards,

Furcy
